time.c

```diff
@@ -2443,6 +2473,142 @@ time_init_args(rb_execution_context_t *ec, VALUE time, VALUE year, VALUE mon, VA
+    size_t ndigits;
+    size_t prec = NIL_P(precision) ? SIZE_MAX : NUM2SIZET(precision);
+
+#define date_time_sep_p(c) ((c) == ' ' || (c) == 'T')
+
+    while ((ptr < end) && ISSPACE(*ptr)) ptr++;
+    year = parse_int(ptr, end, &ptr, &ndigits, true);
+    if (NIL_P(year)) {
+        rb_raise(rb_eArgError, "can't parse: %+"PRIsVALUE, str);
+    }
+    else if (ndigits < 4) {
+        rb_raise(rb_eArgError, "year must be 4 or more digits: %.*s", (int)ndigits, ptr - ndigits);
+    }
+    do {
+#define peekable_p(n) ((ptrdiff_t)(n) < (end - ptr))
+#define peek_n(c, n) (peekable_p(n) && ((unsigned char)ptr[n] == (c)))
+#define peek(c) peek_n(c, 0)
+#define peekc_n(n) (peekable_p(n) ? (int)(unsigned char)ptr[n] : -1)
+#define peekc() peek_n(0)
+#define expect_two_digits(x) (x = two_digits(ptr + 1, end, &ptr, #x))
+        if (!peek('-')) break;
+        expect_two_digits(mon);
+        if (!peek('-')) break;
+        expect_two_digits(mday);
+        if (!peek(' ') && !peek('T')) break;
+        const char *const time_part = ptr + 1;
+        if (!ISDIGIT(peekc_n(1))) break;
+#define nofraction(x) \
+        if (peek('.')) { \
+            rb_raise(rb_eArgError, "fraction " #x " is not supported: %.*s", \
+                     (int)(ptr + 1 - time_part), time_part); \
+        }
+#define need_colon(x) \
+        if (!peek(':')) { \
+            rb_raise(rb_eArgError, "missing " #x " part: %.*s", \
+                     (int)(ptr + 1 - time_part), time_part); \
+        }
+        expect_two_digits(hour);
+        nofraction(hour);
+        need_colon(min);
+        expect_two_digits(min);
+        nofraction(min);
+        need_colon(sec);
+        expect_two_digits(sec);
+        if (peek('.')) {
+            ptr++;
+            for (ndigits = 0; ndigits < prec && ISDIGIT(peekc_n(ndigits)); ++ndigits);
+            if (!ndigits) {
+                int clen = rb_enc_precise_mbclen(ptr, end, rb_enc_get(str));
+                if (clen < 0) clen = 0;
+                rb_raise(rb_eArgError, "subsecond expected after dot: %.*s",
+                         (int)(ptr - time_part) + clen, time_part);
+            }
+            subsec = parse_int(ptr, ptr + ndigits, &ptr, &ndigits, false);
+            if (NIL_P(subsec)) break;
+            while (ptr < end && ISDIGIT(*ptr)) ptr++;
+        }
+    } while (0);
+    while (ptr < end && ISSPACE(*ptr)) ptr++;
+    const char *const zstr = ptr;
+    while (ptr < end && !ISSPACE(*ptr)) ptr++;
+    const char *const zend = ptr;
+    while (ptr < end && ISSPACE(*ptr)) ptr++;
+    if (ptr < end) {
+        VALUE mesg = rb_str_new_cstr("can't parse at: ");
+        rb_str_cat(mesg, ptr, end - ptr);
+        rb_exc_raise(rb_exc_new_str(rb_eArgError, mesg));
+    }
+    if (zend > zstr) {
+        zone = rb_str_subseq(str, zstr - begin, zend - zstr);
+    }
+    if (!NIL_P(subsec)) {
+        /* subseconds is the last using ndigits */
+        if (ndigits < 9) {
```

Comment:
The magic number, 9, relates TIME_SCALE.
It should be replaced with some constant.
TIME_SCALE_NUMDIGITS?